hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -3403,16 +3415,21 @@ private TableGroup consumeEntityJoin(SqmEntityJoin<?> sqmJoin, TableGroup lhsTab
 				sqmJoin.getNavigablePath(),
 				sqmJoin.getExplicitAlias(),
 				null,
-				() -> p -> predicate.set( combinePredicates( predicate.get(), p ) ),
+				() -> p -> needsTreat.setValue( true ),
```

Comment:
I think passing anything other than `null` for the `additionalPredicateCollectorAccess` could become a problem if we are entity joining a multi-table entity, given how the implementation in `AbstractEntityPersister#createRootTableGroup` currently looks like. When `additionalPredicateCollectorAccess` is not `null`, the join predicate for the other tables of the entity will be passed to that collector instead of the `TableReferenceJoin`.

In a query like `select b.superclassAttribute from EntityA a join EntityASub b on a.id = b.id` we might produce wrong SQL because the predicate would be missing for the join from `EntityASub` to it's superclass table.

I don't know a great fix for this yet. Maybe we need to simply ignore `additionalPredicateCollectorAccess` in the table reference join creation. If we can do that, I'd like you to rename that parameter to `discriminatorPredicateCollectorAccess` to make the intent clear.

Reply:
I can't quite follow you on this, I did try ignoring `additionalPredicateCollectorAccess` when creating the table reference join but a test fails (`CustomSQLSecondaryTableTest`).

Not sure this would be a problem for joins though, wouldn't those create `LazyTableGroup`s?

Reply:
Forget that, I misread `AbstractEntityPersister#createRootTableGroup`